python/lsst/daf/butler/formatters/parquet.py

```diff
@@ -21,177 +21,842 @@
+            if par_columns:
+                has_pandas_multi_index = False
+                if b"pandas" in schema.metadata:
+                    md = json.loads(schema.metadata[b"pandas"])
+                    if len(md["column_indexes"]) > 1:
+                        has_pandas_multi_index = True
+
+                if not has_pandas_multi_index:
+                    # Ensure uniqueness, keeping order.
+                    par_columns = list(dict.fromkeys(ensure_iterable(par_columns)))
+                    file_columns = [name for name in schema.names if not name.startswith("__")]
+
+                    for par_column in par_columns:
+                        if par_column not in file_columns:
+                            raise ValueError(
+                                f"Column {par_column} specified in parameters not available in parquet file."
+                            )
+                else:
+                    par_columns = _standardize_multi_index_columns(schema, par_columns)
+
+            filters = self.fileDescriptor.parameters.pop("filters", None)
+
+            if len(self.fileDescriptor.parameters):
+                raise ValueError(
+                    f"Unsupported parameters {self.fileDescriptor.parameters} in ArrowTable read."
+                )
+
+        metadata = schema.metadata if schema.metadata is not None else {}
+        arrow_table = pq.read_table(
+            self.fileDescriptor.location.path,
+            columns=par_columns,
+            use_threads=False,
+            use_pandas_metadata=(b"pandas" in metadata),
+            filters=filters,
+        )
+
+        return arrow_table
+
+    def write(self, inMemoryDataset: Any) -> None:
+        import numpy as np
+        import pandas as pd
+        from astropy.table import Table as astropyTable
+
+        location = self.makeUpdatedLocation(self.fileDescriptor.location)
+
+        if isinstance(inMemoryDataset, pd.DataFrame):
+            arrow_table = pandas_to_arrow(inMemoryDataset)
+        elif isinstance(inMemoryDataset, astropyTable):
+            arrow_table = astropy_to_arrow(inMemoryDataset)
+        elif isinstance(inMemoryDataset, np.ndarray):
+            arrow_table = numpy_to_arrow(inMemoryDataset)
+        elif isinstance(inMemoryDataset, pa.Table):
+            arrow_table = inMemoryDataset
         else:
-            raw_columns = list(self._splitColumnnNames(len(indexes), self.file.metadata.schema.names))
-            self.columns = pd.MultiIndex.from_tuples(raw_columns, names=[f["name"] for f in indexes])
-        self.indexLevelNames = tuple(self.columns.names)
+            raise ValueError("Unsupported type of inMemoryDataset for ParquetFormatter.")
+
+        pq.write_table(arrow_table, location.path)
+
+
+def arrow_to_pandas(arrow_table: pa.Table) -> pd.DataFrame:
+    """Convert a pyarrow table to a pandas DataFrame.
+    If the input table has ``pandas`` metadata in the schema it will be
```

Comment:
Blank line before this one since this is a note and not the short summary. It should probably be attached to the parameter description below.